docs/avansat/persistent-segment-tree.md

```diff
@@ -0,0 +1,236 @@
+
+Așadar, pentru fiecare update, am putea creea o nouă versiune a nodurilor care se schimbă.
+
+Pentru claritate, vom folosi o implementare cu pointeri.
+
+
+```cpp
+struct Node {
+  ll sum;
+  Node* l;
+  Node* r;
+  Node() {
+    sum = 0;
+    l = r = NULL;
+  }
+  Node(ll _sum, Node* _l, Node* _r) {
+    sum = _sum;
+    l = _l;
+    r = _r;
+  }
+};
+
+Node* build(int tl, int tr, int* a) {
+  if (tl == tr) {
+    return new Node {
+      a[tl - 1],
+      NULL, NULL
+    };
+  } else {
+    int mid = (tl + tr) / 2;
+    Node* ret = new Node();
+    ret -> l = build(tl, mid, a);
+    ret -> r = build(mid + 1, tr, a);
+    ret -> sum = ret -> l -> sum + ret -> r -> sum;
+    return ret;
+  }
+}
+
+Node* update(const Node* node, int tl, int tr, int p, int v) {
+  if (tl == tr) {
+    return new Node {
+      v,
+      NULL,
+      NULL
+    };
+  } else {
+    Node *ret = new Node();
+    *ret = *node;
+    int mid = (tl + tr) / 2;
+    if (p <= mid) {
+      ret -> l = update(node -> l, tl, mid, p, v);
+    } else {
+      ret -> r = update(node -> r, mid + 1, tr, p, v);
+    }
+    return ret;
+  }
+}
+
+int n;
+std::vector<Node*> versions;
+
+void init(int _n, int* a) {
+  n = _n;
+  versions.push_back(build(1, n, a));
+}
+
+void updateValue(int pos, int value) {
+  versions.push_back(update(versions.back(), 1, n, pos, value)); // creem o versiune nouă, pornim inițial cu versiunea trecută
+}
+
+ll sequenceQuery(int l, int r, int t) {
+  return query(versions[t], 1, n, l, r);
+}
+```
+
+!!! info "Atenție!"
+    Arborele de interval persistent nu poate fi folosit cu lazy propagation, deoarece nu știm pentru fiecare nod când se modifică.
+
+# Problema History
+
+Această problemă este foarte asemănătoare cu problema precedentă, dar în loc să răspundem la interogări de forma "care e suma valorilor din această subsecvență?", treubie să răspundem la întrebări de forma "care e suma maximă a unei subsecvențe incluse în această subsecvență?".
+
+Știm că putem răspunde la înterbări de acest fel cu un arbore de intervale puțin modificat. Acum, tot ce trebuie e să îl facem persistent.
+
+```cpp
+#include <algorithm>
+#include <vector>
+
+int readNumber();
+
+struct Node {
+  long long sum, value, prefmax, suffmax;
+  Node* l;
+  Node* r;
+  Node() {}
+  Node(int x) {
+    sum = value = prefmax = suffmax = x;
+    l = r = NULL;
+  }
+  Node operator + (const Node &other) const {
+    Node ret;
+    ret.sum = sum + other.sum;
+    ret.prefmax = std::max(prefmax, sum + other.prefmax);
+    ret.suffmax = std::max(other.suffmax, other.sum + suffmax);
+    ret.value = std::max({value, other.value, suffmax + other.prefmax, ret.prefmax, ret.suffmax});
+    return ret;
+  }
+  void refresh() {
+    Node* ll = l;
+    Node* rr = r;
+    *this = *l + *r;
+    this -> l = ll;
+    this -> r = rr;
+  }
```

Comment:
Ai cam tras asta de păr, if you ask me. Era de ajuns:
```cpp
    void refresh() {
        if (l && r) {
            *this = *l + *r;
        }
    }
```
că ai deja `operator+` (cel puțin așa am interpretat codul tău).

Nu e nevoie să scrii `this->` dacă ai referință la un membru sau funcție, se subînțelege și scrii aia doar dacă ai conflict de nume.

Reply:
Asta nu cred ca merge ca in functia de join nu imi fac si fiii si atunci this->l (si this->r) o sa fie nullptr daca fac asta

Reply:
Păi și atunci tu de ce dereferențiezi ceva în codul tău ce va fi potențial nullptr? Eu m-am asigurat că ambele-s referințe valide (nullptr e fals), tu nu.